File: flutter/test/sellwild_config_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sellwild_sdk/sellwild_sdk.dart';

import 'factories/shape_factories.dart';

void main() {
  group('SellwildConfig', () {
    test('has correct defaults', () {
      const config = SellwildConfig(
        partnerCode: 'test',
        listingsUrl: 'https://cache.sellwild.com/listings-img-data-sm',
      );

      expect(config.partnerCode, 'test');
      expect(config.titleSize, 16);
      expect(config.fontSize, 13);
      expect(config.boltive, false);
      expect(config.lotame, false);
      expect(config.debug, false);
      expect(config.adRefreshInterval, const Duration(seconds: 30));
      expect(config.hideBannerTop, false);
    });

    test('toJson includes partnerCode and listingsUrl', () {
      const config = SellwildConfig(
        partnerCode: 'mypartner',
        listingsUrl: 'https://cache.sellwild.com/listings-img-data-sm',
      );

      final json = config.toJson();
      expect(json['partnerCode'], 'mypartner');
      expect(json['listingsUrl'], 'https://cache.sellwild.com/listings-img-data-sm');
      expect(json.containsKey('title'), isFalse);
      expect(json['linkText'], 'View all');
    });

    test('toJson includes optional values only when set', () {
      const config = SellwildConfig(
        partnerCode: 'mypartner',
        title: 'Deals',
        linkText: null,
        buyNowText: null,
        gamTag: '/1234/sellwild',
        gptProxyUrl: 'https://gpt.example.com/gpt.js',
      );

      final json = config.toJson();
      expect(json['title'], 'Deals');
      expect(json['gamTag'], '/1234/sellwild');
      expect(json['gptProxyUrl'], 'https://gpt.example.com/gpt.js');
      expect(json.containsKey('linkText'), isFalse);
      expect(json.containsKey('buyNowText'), isFalse);
    });

    test('kill switches default on, rate 1', () {
      const config = SellwildConfig(partnerCode: 'test');

      expect(config.eventsEnabled, isTrue);
      expect(config.failuresEnabled, isTrue);
      expect(config.failuresSampleRate, 1.0);
    });
  });

  test('PrebidServerConfig keeps its values and defaults', () {
    // Built from a run-time list (not const) so the constructor is measured.
    final bidders = ['appnexus', 'rubicon'].toList();
    final server = PrebidServerConfig(
      accountId: 'acct',
      endpoint: 'https://pbs.example.com/openrtb2/auction',
      bidders: bidders,
    );

    expect(server.accountId, 'acct');
    expect(server.bidders, bidders);
    expect(server.timeout, 1500);
    expect(server.syncEndpoint, isNull);
  });

  group('SellwildAdSize', () {
    test('banner dimensions', () {
      expect(SellwildAdSize.banner320x50.width, 320);
      expect(SellwildAdSize.banner320x50.height, 50);
      expect(SellwildAdSize.mrec300x250.width, 300);
      expect(SellwildAdSize.mrec300x250.height, 250);
    });

    test('label is WIDTHxHEIGHT', () {
      expect(SellwildAdSize.values.map((s) => s.label),
          ['320x50', '300x250', '728x90', '300x600', '160x600']);
    });
  });

  group('SellwildListing', () {
    final listings = ListingFactory();

    test('parses from JSON', () {
      final json = listings.build();

      final listing = SellwildListing.fromJson(json);

      expect(listing.id, json['id']);
      expect(listing.title, json['title']);
      expect(listing.displayPrice, json['price']);
      expect(listing.primaryPhoto?.url, (json['photos'] as List).first['url']);
    });

    test('displayPrice rounds a fractional price', () {
      final listing = SellwildListing.fromJson(listings.fractionalPrice());
      expect(listing.displayPrice, '50');
    });

    test('displayPrice returns null for zero price', () {
      final listing = SellwildListing.fromJson(listings.zeroPrice());
      expect(listing.displayPrice, null);
    });
  });
}

File: android/src/test/kotlin/com/sellwild/sdk/SellwildConfigTest.kt
package com.sellwild.sdk

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** SellwildConfig defaults, its JSON form, and the other public config types. */
class SellwildConfigTest {

    @Test
    fun `default config has expected values`() {
        val config = SellwildConfig(partnerCode = "test_partner", listingsUrl = "https://cache.sellwild.com/listings-img-data-sm")

        assertEquals("test_partner", config.partnerCode)
        assertEquals(16, config.titleSize)
        assertEquals(13, config.fontSize)
        assertFalse(config.boltive)
        assertFalse(config.lotame)
        assertFalse(config.debug)
        assertEquals(30_000L, config.adRefreshIntervalMs)
        assertFalse(config.hideBannerTop)
        assertFalse(config.hideBannerBottom)
    }

    @Test
    fun `toJson writes every field, and the optional ones only when set`() {
        val full = SellwildConfig(
            partnerCode = "mypartner",
            title = "Deals",
            gamTag = "/1234/fixture",
            gptProxyUrl = "https://gpt.invalid/proxy",
            boltive = true,
            boltiveClientId = "antengo",
            debug = true,
        ).toJson()
        val bare = SellwildConfig(partnerCode = "mypartner", linkText = null, buyNowText = null).toJson()

        assertEquals("mypartner", full.getString("partnerCode"))
        assertEquals(SellwildConfig.DEFAULT_LISTINGS_URL, full.getString("listingsUrl"))
        assertEquals("Deals", full.getString("title"))
        assertEquals("View all", full.getString("linkText"))
        assertEquals("Buy now", full.getString("buyNowText"))
        assertEquals("/1234/fixture", full.getString("gamTag"))
        assertEquals("https://gpt.invalid/proxy", full.getString("gptProxyUrl"))
        assertTrue(full.getBoolean("boltive"))
        assertEquals("antengo", full.getString("boltiveClientId"))
        assertTrue(full.getBoolean("debug"))
        assertEquals(30_000L, full.getLong("adRefreshInterval"))
        for (key in listOf("title", "linkText", "buyNowText", "gamTag", "gptProxyUrl")) assertFalse(key, bare.has(key))
        assertEquals(25, bare.length())
    }

    @Test
    fun `an empty listings URL falls back to the general cache`() {
        assertEquals(SellwildConfig.DEFAULT_LISTINGS_URL, SellwildConfig(partnerCode = "p", listingsUrl = "").effectiveListingsUrl)
        assertEquals(SellwildConfig.DEFAULT_LISTINGS_URL, SellwildConfig(partnerCode = "p").effectiveListingsUrl)
        assertEquals("https://c.invalid/x", SellwildConfig(partnerCode = "p", listingsUrl = "https://c.invalid/x").effectiveListingsUrl)
    }

    @Test
    fun `AdSize has the IAB dimensions and a WxH label`() {
        assertEquals(
            listOf("320x50", "300x250", "728x90", "300x600", "160x600"),
            AdSize.values().map { it.label },
        )
        assertEquals(320, AdSize.BANNER_320x50.width)
        assertEquals(250, AdSize.MREC_300x250.height)
    }

    @Suppress("DEPRECATION")
    @Test
    fun `the settings types keep what they are given`() {
        val growth = SellwildGrowthCodeConfig(true, "pid", "https://gc.invalid", "weatherbug.com", false, 12)
        val localized = SellwildLocalizedListingsConfig(true, "sportserver", "https://c.invalid/", "x-{state}.json", 25, "AL")
        val server = PrebidServerConfig("acct", "https://pbs.invalid/openrtb2/auction", listOf("appnexus"), 900, "https://pbs.invalid/cookie_sync")
        val ix = IxConfig(siteIdM = "m", siteIdD = "d")
        val openx = OpenxConfig(delDomain = "x.openx.net", unitM = "1", unitD = "2")
        val pubmatic = PubmaticConfig(pubIdM = "p", adSlotM = "m", adSlotD = "d")
        val appnexus = AppnexusConfig(placementIdM = 1, placementIdD = 2)
        val waterfall = WaterfallPartnerConfig(false, 0.1f, 0.2f, "a", "b", "c", "d", 0.5f, 0.6f, 3, 60_000L, "US")

        assertEquals(12, growth.ttlHours)
        assertEquals("AL", localized.forceState)
        assertEquals(1500, PrebidServerConfig("acct", "https://pbs.invalid", emptyList()).timeout)
        assertEquals(900, server.timeout)
        assertFalse(ix.disabled || openx.disabled || pubmatic.disabled || appnexus.disabled || waterfall.disabled)
        assertEquals("US", waterfall.geo)
        assertEquals(
            SellwildConfig(partnerCode = "p", growthCode = growth, localizedListings = localized, prebidServer = server, ix = ix),
            SellwildConfig(partnerCode = "p").copy(growthCode = growth, localizedListings = localized, prebidServer = server, ix = ix),
        )
    }
}
